source/extensions/filters/http/grpc_http1_bridge/http1_bridge_filter.cc

```diff
@@ -67,35 +78,65 @@ Http::FilterTrailersStatus Http1BridgeFilter::encodeTrailers(Http::ResponseTrail
   }
 
   if (do_bridging_) {
-    // Here we check for grpc-status. If it's not zero, we change the response code. We assume
-    // that if a reset comes in and we disconnect the HTTP/1.1 client it will raise some type
-    // of exception/error that the response was not complete.
-    const Http::HeaderEntry* grpc_status_header = trailers.GrpcStatus();
-    if (grpc_status_header) {
-      uint64_t grpc_status_code;
-      if (!absl::SimpleAtoi(grpc_status_header->value().getStringView(), &grpc_status_code) ||
-          grpc_status_code != 0) {
-        response_headers_->setStatus(enumToInt(Http::Code::ServiceUnavailable));
-      }
-      response_headers_->setGrpcStatus(grpc_status_header->value().getStringView());
-    }
-
-    const Http::HeaderEntry* grpc_message_header = trailers.GrpcMessage();
-    if (grpc_message_header) {
-      response_headers_->setGrpcMessage(grpc_message_header->value().getStringView());
-    }
-
-    // Since we are buffering, set content-length so that HTTP/1.1 callers can better determine
-    // if this is a complete response.
-    response_headers_->setContentLength(
-        encoder_callbacks_->encodingBuffer() ? encoder_callbacks_->encodingBuffer()->length() : 0);
+    // We're bridging, so we need to process trailers and set the http status, content length,
+    // grpc status, and grpc message from those trailers.
+    doResponseTrailers(trailers);
   }
 
   // NOTE: We will still write the trailers, but the HTTP/1.1 codec will just eat them and end
   //       the chunk encoded response which is what we want.
   return Http::FilterTrailersStatus::Continue;
 }
 
+void Http1BridgeFilter::updateHttpStatusAndContentLength(
+    const Http::ResponseHeaderOrTrailerMap& trailers,
+    bool enable_http_status_codes_in_trailers_response) {
+  // Here we check for grpc-status. If it's not zero, we change the response code. We assume
+  // that if a reset comes in and we disconnect the HTTP/1.1 client it will raise some type
+  // of exception/error that the response was not complete.
+  const Http::HeaderEntry* grpc_status_header = trailers.GrpcStatus();
+  if (grpc_status_header) {
+    uint64_t grpc_status_code;
+    if (enable_http_status_codes_in_trailers_response &&
+        (!absl::SimpleAtoi(grpc_status_header->value().getStringView(), &grpc_status_code) ||
+         grpc_status_code != 0)) {
+      response_headers_->setStatus(Grpc::Utility::grpcToHttpStatus(grpc_status_code));
+    }
+  }
+
+  // Since we are buffering, set content-length so that HTTP/1.1 callers can better determine
```

Comment:
@snowp, do you mean we should still support initial 503 behavior for headers response even though a gRPC mapping would be more accurate?

Reply:
I just mean that if we're going to respond with a headers only response before seeing upstream data we won't be buffering, so this comment seems like it's not right in all cases that we use this helper function? I think either rewording the comment to talk about "we may be buffering" or changing it so that we only call the code that updates the content-length when we are actually buffering. If we need this piece of code even when we are not buffering, I think just updating the comment would be good enough